### PR TITLE
Resolve symlinks more in nspire-* scripts

### PR DESCRIPTION
**Pull request by coolreader18**
Related to #208

### CODE REVIEW

**Comment by coolreader18** [comment]
Could someone review this?

**Comment by adriweb** [comment]
Haven't tested but looks OK to me.
But anyway this is for @Vogtinator to review/merge.

**Comment by lights0123** [comment]
While we're at it, can we make the ARM toolchain relative too? It would be nice to be able to always have `nspire-gcc` in my path without causing problems when trying to use normal arm-none-eabi-gcc for any other ARM platform, without resorting to [special](https://github.com/lights0123/ndless-rust/blob/master/src/Installing/Linux.md#pro-tip) [environment variables](https://github.com/lights0123/ndless-rs/blob/866ed9d816c12b14ef017476212878e1ddce1d1e/cargo-ndless/src/main.rs#L34).

**Comment by Vogtinator** [comment]
For that, it would be necessary to drop the possibility of using a system toolchain. I don't think that ever actually worked though (needs gcc built with newlib).

**Comment by lights0123** [comment]
Yeah. Compiling applications with standard ARM gcc or LLVM works just fine, but linking with them causes lots of problems. I tried to do something with LLVM's lld (mainly because it's faster than ld.gold and the like), but I was getting internal errors. If we can get it to work that's great, but there's no reason to have the standard installation be worse because it supports something that doesn't work anyways.

**Comment by coolreader18** [comment]
@Vogtinator I've updated all the scripts to use `DIRNAME=$(nspire-tools path)`, I think they're much clearer now.

**Comment by coolreader18** [comment]
An easy way to support a relative toolchain without completely giving up a system toolchain would be to just add `"$(nspire-tools path)/toolchain/install/bin` to PATH in all the scripts, so that if the sdk doe ever use the system toolchain it'll just find it in PATH.

File: ndless-sdk/bin/nspire-tools [hunk 1]
```diff
@@ -12,15 +12,13 @@ DIRNAME="$(dirname "$0")"
...
-    SYMLINK="$(readlink "$0")"
-    
-    # On MacOS X, the readlink command hasn't the same behavior than the others operating systems based on Unix.
-    if [ "$UNAME" == "darwin" ]; then
-    	DIRNAME="$(dirname "$SYMLINK")"
-    else
-    	DIRNAME="${SYMLINK}/.."
-    fi
+{
+  (which realpath && alias realpath=realpath) || (readlink -f / && alias realpath='readlink -f') || \
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:16]: This is really hard to read, can it be converted into more straightforward `if foo; then bar; fi` syntax?

**Comment by Vogtinator** [comment]
> An easy way to support a relative toolchain without completely giving up a system toolchain would be to just add `"$(nspire-tools path)/toolchain/install/bin` to PATH in all the scripts, so that if the sdk doe ever use the system toolchain it'll just find it in PATH.

Agreed. Originally it was supposed to support a system provided ARM toolchain as well, but that's been broken for quite some time now, with the introduction of newlib. So if you do that, you could also drop the `$(which arm-none-eabi-gcc ...)` calls.

**Comment by coolreader18** [comment]
I didn't do exactly what I described in my previous comment, but this way it should be easier to patch for packaging.

File: ndless-sdk/bin/nspire-tools [hunk 2]
```diff
@@ -61,7 +81,17 @@ case "$1" in
 		cp "${DIRNAME}/main.c.tpl" main.c
 		;;
 	path)
-		echo "$(dirname "$DIRNAME")/"
+		echo "$NDLESS/"
+		;;
+	gcc)
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:86]: This seems odd, why only `gcc` and not other tools?

IMO you can remove this, as with `toolchainbin` it's redundant.

File: ndless-sdk/bin/nspire-tools [hunk 3]
```diff
@@ -2,41 +2,61 @@
...
 
 if [ $# -eq 0 ]; then
 	usage
 	exit 0
 fi
 
-case "$1" in
+# following functions can be patched when packaging/distributing
+toolchainbin() {
+	echo "$NDLESS/toolchain/install/bin"
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:46]: An idea: Make this print an environment variable instead if set, so that it can be overwritten simply by doing an `export`.

Patching should ideally not be necessary.

File: ndless-sdk/bin/nspire-tools [hunk 4]
```diff
@@ -61,7 +81,17 @@ case "$1" in
...
+		echo "$NDLESS/"
+		;;
+	gcc)
+		GCC=$(toolchainbin)/arm-none-eabi-gcc
+		exec "$GCC" "$@"
+		;;
+	toolchainbin)
+		toolchainbin
+		;;
+	_zehnpath)
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:93]: `_zehn_loader_path`

File: ndless-sdk/bin/arm-none-eabi-ld.gold [hunk 5]
```diff
@@ -33,16 +21,18 @@ done
...
 fi
 
-(cd "$DIRNAME/../system" && MAKEFLAGS= make -s all)
+MAKEFLAGS='' make -C "$NDLESS/system" -s all
 
 home="${USERPROFILE:-$HOME}"
 mkdir -p "$home/.ndless/lib"
 
-$(basename "$0" | cut -f1 -d'.') --pic-veneer --emit-relocs -T "${DIRNAME}/../system/ldscript" -static "${DIRNAME}/../system/crt0.o" "${DIRNAME}/../system/crti.o" "${args[@]}" "${stdlib[@]}" "${DIRNAME}/../system/crtn.o" -L "${DIRNAME}/../lib" -L "${home}/.ndless/lib"
+LD=$(nspire-tools toolchainbin)/arm-none-eabi-ld
```
Comment (Vogtinator) [inline ndless-sdk/bin/arm-none-eabi-ld.gold:36]: Needs quoting

File: ndless-sdk/bin/nspire-g++ [hunk 6]
```diff
@@ -1,20 +1,8 @@
...
-
-if [ "$UNAME" = "cygwin" ]; then
-	DIRNAME="$(cygpath --path --windows "$DIRNAME")"
-fi
-
-GCC="$( (which arm-elf-g++ arm-none-eabi-g++ arm-linux-gnueabi-g++ | head -1) 2>/dev/null)"
 home="${USERPROFILE:-$HOME}"
 mkdir -p "$home/.ndless/include"
-"$GCC" -mcpu=arm926ej-s -fno-use-cxa-atexit -D _TINSPIRE -fuse-ld=gold "$@" -I "$home/.ndless/include" -I "${DIRNAME}/../include" -I "${DIRNAME}/../include/freetype2"
+exec nspire-tools gcc -mcpu=arm926ej-s -fno-use-cxa-atexit -D _TINSPIRE -fuse-ld=gold "$@" -I "$home/.ndless/include" -I "${NDLESS}/include" -I "${NDLESS}/include/freetype2"
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-g++:8]: `"$(nspire-tools toolchainbin)/arm-none-eabi-g++"`

File: ndless-sdk/bin/nspire-tools [hunk 7]
```diff
@@ -61,7 +81,17 @@ case "$1" in
...
 		;;
 	path)
-		echo "$(dirname "$DIRNAME")/"
+		echo "$NDLESS/"
+		;;
+	gcc)
+		GCC=$(toolchainbin)/arm-none-eabi-gcc
+		exec "$GCC" "$@"
+		;;
+	toolchainbin)
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:90]: I'd prefer `toolchainpath` for symmetry.

Additionally, it's "private" like `_zehnpath`, so `_toolchainpath`

File: ndless-sdk/bin/nspire-tools [hunk 8]
```diff
@@ -2,41 +2,61 @@
...
 
-case "$1" in
+# following functions can be patched when packaging/distributing
+toolchainbin() {
+	echo "$NDLESS/toolchain/install/bin"
+}
+zehn_loader_path() {
+	MAKEFLAGS='' make -C "${NDLESS}/tools/zehn_loader" -s all
+	local suffix=$1
+	echo "$NDLESS/tools/zehn_loader/zehn_loader$suffix.tns"
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:51]:
```suggestion
	echo "$NDLESS/tools/zehn_loader/zehn_loader${suffix}.tns"
```

File: ndless-sdk/bin/nspire-tools [hunk 9]
```diff
@@ -2,41 +2,61 @@
...
-    # On MacOS X, the readlink command hasn't the same behavior than the others operating systems based on Unix.
-    if [ "$UNAME" == "darwin" ]; then
-    	DIRNAME="$(dirname "$SYMLINK")"
-    else
-    	DIRNAME="${SYMLINK}/.."
-    fi
-}
+{
+  if which realpath; then
+    alias realpath=realpath
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:19]: Why make an alias for something which already exists?

File: ndless-sdk/bin/nspire-tools [hunk 10]
```diff
@@ -2,41 +2,61 @@
...
+  elif readlink -f /; then
+    alias realpath='readlink -f'
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:21]: The alias is only used once, so could just do the call directly

File: ndless-sdk/bin/nspire-tools (hunk 3 above)
Comment (coolreader18) [reply ndless-sdk/bin/nspire-tools:46]: I think patching is kind of unavoidable in some cases; requiring a user to set an environment variable after they install a package isn't very ergonomic, and different distros have different ways of organizing packages. Plus, a patch for these functions is pretty minimal, my aur patch for this section is just:

```patch
 # following functions can be patched when packaging/distributing
 toolchainbin() {
-	echo "$NDLESS/toolchain/install/bin"
+	echo "$NDLESS/bin"
 }
 zehn_loader_path() {
-	MAKEFLAGS='' make -C "${NDLESS}/tools/zehn_loader" -s all
 	local suffix=$1
-	echo "$NDLESS/tools/zehn_loader/zehn_loader$suffix.tns"
+	echo "$NDLESS/lib/zehn_loader$suffix.tns"
 }
```

Comment (Vogtinator) [reply ndless-sdk/bin/nspire-tools:46]: Whoops, accidentially marked as resolved...

I think that a patch is avoidable in this case. You can do that in a variety of ways, like adding a file into `/etc/profile.d/` or placing a wrapper into `/usr/bin/` which just sets the variable.

It should work just fine to use a package provided toolchain with the latest SDK checkout from git.

Comment (coolreader18) [reply ndless-sdk/bin/nspire-tools:43]: What about with `zehn_loader_path`? It still needs to be able to pass whether it's compressed to the function; how would that work with an environment variable?

Comment (Vogtinator) [reply ndless-sdk/bin/nspire-tools:43]: If `zehn_loader_path` only returns the path to the precompiled binaries, that should work fine.

File: ndless-sdk/bin/nspire-g++ [hunk 11]
```diff
@@ -1,20 +1,9 @@
...
-
-if [ "$UNAME" = "cygwin" ]; then
-	DIRNAME="$(cygpath --path --windows "$DIRNAME")"
-fi
-
-GCC="$( (which arm-elf-g++ arm-none-eabi-g++ arm-linux-gnueabi-g++ | head -1) 2>/dev/null)"
 home="${USERPROFILE:-$HOME}"
 mkdir -p "$home/.ndless/include"
-"$GCC" -mcpu=arm926ej-s -fno-use-cxa-atexit -D _TINSPIRE -fuse-ld=gold "$@" -I "$home/.ndless/include" -I "${DIRNAME}/../include" -I "${DIRNAME}/../include/freetype2"
+G++="$(nspire-tools _toolchainpath)/arm-none-eabi-g++"
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-g++:8]: `G++` is not an allowed variable name, this will fail to run.

File: ndless-sdk/bin/nspire-tools [hunk 12]
```diff
@@ -2,41 +2,58 @@
...
+  else
+    real0=$0
+  fi
+
+  DIRNAME="$(dirname "$real0")"
+} >/dev/null 2>&1
+
 
 # Cygwin owns a dedicated command to find the symlink target
 if [ "$UNAME" = "cygwin" ]; then
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:31]: Not really part of this PR, but I wonder whether this shouldn't be part of the ladder above?
Otherwise it would try to resolve the symlink target of the parent dir. I assume it was expected that `readlink` doesn't work on cygwin?

I guess noone uses cygwin anymore anyway though.

File: ndless-sdk/bin/nspire-tools [hunk 13]
```diff
@@ -2,41 +2,58 @@
...
 	DIRNAME="$(cygpath --path --windows "$DIRNAME")"
 fi
+NDLESS="$(dirname "$DIRNAME")"
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:34]: Is actually the path to the SDK, so `SDKROOT`?

File: ndless-sdk/bin/nspire-as [hunk 14]
```diff
@@ -1,25 +1,7 @@
...
-
-if [ "$UNAME" = "cygwin" ]; then
-  DIRNAME="$(cygpath --path --windows "$DIRNAME")"
-fi
+NDLESS="$(nspire-tools path)"
 
 # gcc instead of as, to call the C preprocessor
-GCC=$( (which arm-elf-gcc arm-none-eabi-gcc arm-linux-gnueabi-gcc | head -1) 2>/dev/null)
-"$GCC" -mcpu=arm926ej-s -D GNU_AS "$@" -I "$DIRNAME/../include"
+exec nspire-tools gcc -mcpu=arm926ej-s -D GNU_AS "$@" -I "$NDLESS/include"
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-as:7]:
```suggestion
exec "$(nspire-tools _toolchainpath)/arm-none-eabi-gcc" -mcpu=arm926ej-s -D GNU_AS "$@" -I "$NDLESS/include"
```

File: ndless-sdk/bin/nspire-tools (hunk 12 above)
Comment (coolreader18) [reply ndless-sdk/bin/nspire-tools:31]: I'm not sure exactly how it would interact, but I don't know if unix `dirname` would work on a path output by `cygpath`, so that's why I left it in that order; it still does `dirname $real0` after the ladder.

Comment (Vogtinator) [reply ndless-sdk/bin/nspire-tools:31]: Ne neither, but it already calls `dirname` on the output anyway, so that probably worked at some point.

Comment (coolreader18) [reply ndless-sdk/bin/nspire-tools:31]: Oh, true. Maybe `SDKROOT=...` should be before the dirname? Idk, as you said, no one really uses cygwin anyway.

File: ndless-sdk/bin/nspire-tools (hunk 3 above)
Comment (coolreader18) [reply ndless-sdk/bin/nspire-tools]: I did basically that, `$_NDLESS_ZEHN_PATH` has the `zehn_loader[_compressed].tns` files.

File: ndless-sdk/bin/nspire-as (hunk 14 above)
Comment (Vogtinator) [reply ndless-sdk/bin/nspire-as:8]: Still uses `nspire-tools gcc`

File: ndless-sdk/bin/nspire-tools (hunk 12 above)
Comment (Vogtinator) [reply ndless-sdk/bin/nspire-tools:31]: Not really, as `$DIRNAME` is only used for determining `$SDKROOT`.

According to https://cygwin.com/cygwin-ug-net/cygpath.html, `--windows` means that it returns a Windows path, with `--path` being an unrelated and apparently wrong option here? It doesn't say anything about link resolution...
Assuming that the comment is correct, the intention was probably

```
elif which readlink; then
    real0="$(readlink "$0")" || real0=$0
elif  [ "$UNAME" = "cygwin" ]; then
    real0="$(cygpath "$0")" || real0=$0
else
    real0=$0
fi
```

Do you agree? Though I'd be in favor of just dropping the cygwin case here.

Comment (coolreader18) [reply ndless-sdk/bin/nspire-tools:31]: Yeah, removing the cygwin code path might be best, it's incorrect and nobody's opened an issue about it, and wsl or msys2 is probably more common these days.

File: ndless-sdk/bin/nspire-as (hunk 14 above)
Comment (coolreader18) [reply ndless-sdk/bin/nspire-as:8]: Oh, whoops :sweat_smile:

File: ndless-sdk/bin/nspire-tools [hunk 15]
```diff
@@ -2,41 +2,43 @@
 
 function usage() {
 	echo "Usage: nspire-tools <command> <args>"
-	echo -e "   new <program>:   Generate a standard Makefile for <program>.tns"
-	echo -e "   main:            Create an empty main.c"
+	echo "   new <program>:   Generate a standard Makefile for <program>.tns"
+	echo "   main:            Create an empty main.c"
+	echo "   path:            Print out the path of the ndless installation directory"
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:7]: A bit misleading and arguably wrong:

```suggestion
	echo "   path:            Print out the path of the ndless SDK directory"
```

File: ndless-sdk/bin/nspire-tools [hunk 16]
```diff
@@ -61,7 +63,21 @@ case "$1" in
 		cp "${DIRNAME}/main.c.tpl" main.c
 		;;
 	path)
-		echo "$(dirname "$DIRNAME")/"
+		echo "$SDKROOT/"
+		;;
+	_toolchainpath)
+		if [[ -n $_NDLESS_TOOLCHAIN_PATH ]]; then
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:69]:
```suggestion
		if [[ -z $_NDLESS_TOOLCHAIN_PATH ]]; then
```

File: ndless-sdk/bin/nspire-tools [hunk 17]
```diff
@@ -61,7 +63,21 @@ case "$1" in
...
+			_NDLESS_TOOLCHAIN_PATH="$SDKROOT/toolchain/install/bin"
+		fi
+		echo "$_NDLESS_TOOLCHAIN_PATH"
+		;;
+	_zehn_loader_path)
+		if [[ -n $_NDLESS_ZEHN_PATH ]]; then
```
Comment (Vogtinator) [inline ndless-sdk/bin/nspire-tools:75]:
```suggestion
		if [[ -z $_NDLESS_ZEHN_PATH ]]; then
```
